### ISSUE TITLE
ratelimit on cluster scale-up

### ISSUE THREAD

**Issue by sulixu** [kubernetes/autoscaler#3916]
**Which component are you using?**: cluster autoscaler 1.16.7

**Is your feature request designed to solve a problem? If so describe the problem this feature should solve.**:
I would like have a sort of ratelimit mechanism on cluster scale-up, for example if cluster autoscaler decides to scale up 500 nodes at a time because of a submission from 1 batch workload, we don't want to scale up 500 nodes all at once, it would be nice to have a few controlled and gradual scale-ups to alleviate the sudden load to kube api server. 

**Describe the solution you'd like.**:

I have a draft on https://github.com/sulixu/autoscaler/commit/3e53de7caeb1002fe1e8a6e7026ebaaf632222cf and tested the scale-up ratelimiting, it seems to be working as expected.

for example I set 5 nodes per batch per min (with burst at 10 nodes per min)

to scale up 25 nodes: (with 40 secs scan-interval)

t1 -> t2 -> t3 -> t4 -> t5 -> t6 -> t7
10 -> idle -> 5 -> idle -> 5 -> idle -> 5

I understand this ratelimiting won't be fair for all clients, for example at the same time 1 batch workload has 500 pods pending, and 1 long running service have 2 pods pending, if all the pending pods have the same priority, the 2 pods from the long running services will always have a lower chance of getting scheduled, and will likely be pending until most of batch pods have scheduled. however we would set the pod priority to let the long running service's pod always have a higher priority than others.

**Describe any alternative solutions you've considered.**:

increasing the scan-interval won't help as it cannot control the number of new nodes in a single scale-up

**Additional context.**:

We have seen that a sudden massive increase in nodes during cluster scale-up can overwhelm the cluster. the control plane got into a busy state and it overloaded the api server.

**Comment by sonowz** [comment]
> We have seen that a sudden massive increase in nodes during cluster scale-up can overwhelm the cluster. the control plane got into a busy state and it overloaded the api server.

I second this

**Comment by MaciekPytel** [comment]
For anyone looking to implement this - I think it could probably be implemented fairly easily as a NodeGroupSetProcessor (one that rate limits scale-up rather than just splitting it across zones).

**Comment by anson627** [comment]
Hi @MaciekPytel , can you elaborate on how to use NodeGroupSetProcessor to implement rate limits scale-up? Also I'm exploring ways to circuit break scale up when cluster reaching certain size (in terms of # of pods and # of nodes)

**Comment by innovate-invent** [comment]
I am looking for very similar functionality, but I wasn't happy with the idea of an global scale up rate limit.

I was thinking it would be better to allow tagging resources with a "max-age-before-scaling" tag. The autoscaler would look at the tag value, compare it to the resource creation time, and only include it in the scaling decision if the max age has been exceeded.
Another tag "max-scale-candidates" would exclude the resource if the number of pods requesting resources has already exceeded its value.

Combined this will allow some resources to lower their "priority" when considering if resources should be scaled up. This also gives the application creating the resources some control over how the scaler behaves at runtime.

**Comment by damirda** [comment]
@innovate-invent: I like the idea of setting `max-age-before-scaling` per pod, and it should be easy to implement, but what do you mean by "tagging pods"? You mean by using annotations?

**Comment by innovate-invent** [comment]
yes, annotations. I need a way to specify these per pod.

**Comment by x13n** [comment]
Making this a per-pod setting complicates the logic. Can you describe the problem you're trying to solve by making it per-pod rather than a global flag?

Unless I'm missing something, `max-age-before-scaling` annotation doesn't solve the problem of overloading kube-apiserver. If a 500-replica batch job is created, it will still do a single huge scale up, except it will happen `max-age-before-scaling` later than it normally would without this annotation.

Re: `max-scale-candidates` - what does it mean to have this on the pod level? What should CA do if there are multiple pods with different values in this annotation?

**Comment by innovate-invent** [comment]
I didn't have batch jobs in mind when describing this, but I think it would still work.
Lets say we submit a 500 replica batch job with annotations `max-age-before-scaling: 20s`, `max-scale-candidates: 10`, and `parallelism: 50`.
The scaler collects a complete list of all unassigned pods. The list is sorted ascending on `max-age-before-scaling` (with a default of 0 if unspecified) and pods are assigned to available nodes until the nodes are saturated. The remaining list is then truncated on `max-age-before-scaling < create-time - now()`. The truncated list is then sorted descending on the `max-scale-candidates` annotation (defaulting to inf when unspecified) and truncated when list_index > `max-scale-candidates`.

This assumes that a pods create time is when the number of parallel pods goes below the `parallelism` value and not the create time of the entire batch job. 

The batch job will then execute as many pods as it can in 20s before scaling the cluster up to accommodate an additional 10 pods in parallel. This will repeat at the scalers scan interval until all jobs are done, or the cluster has scaled to its maximum, or is large enough to accommodate 50 pods in parallel.

We may also want to include an optional `max-scale-group` annotation so that we can provide a mechanism to prevent multiple batch jobs competing for the `max-scale-candidates` pool.

The reason I want this annotation to be applied per pod/job/replicaset is because I want each job to have different scaling priorities depending on the task it is completing. ie, a low priority job might have much much larger `max-age-before-scaling` or smaller `max-scale-candidates` than other jobs.

**Comment by x13n** [comment]
Note that CA cannot guarantee which pods will get scheduled on the newly-provisioned nodes. So, in this scenario, a large, but low priority batch job can eat up the resources even though it was another job with higher scaling priority that triggered the scale up.

**Comment by innovate-invent** [comment]
Right, that is too easy to conflate. My description still (somewhat) holds when evaluating scaling priority, what actually happens between the CA intervals is up to the scheduler. If a higher priority job has a bunch of tasks that age out in that time, then it would potentially scale up a larger amount on the next interval (given a permissive `max-scale-candidates`).

This all could interplay with [Non-preempting PriorityClass](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#non-preempting-priority-class) and scheduler back-off, allowing for something much closer to what I described.

**Comment by x13n** [comment]
Right, combining CA settings with scheduler settings should then allow the behavior you want (although the exact config might get a bit complicated).

What I'm not clear on though is **why** you want this in the first place. The concern in the original post was around helping to keep the control plane stable. Global setting is enough to achieve this. What are you trying to achieve by having certain workloads cause slower scale ups than others? The cost shouldn't be a factor - a batch job will eat the same total amount of core-hours when run for N hours on 1 machine and when run for 1h on N machines. And if you have a higher priority job, it can simply evict the low priority one. So what is the use case here?

**Comment by innovate-invent** [comment]
The load to kube api server isn't so much my concern than the fact that a burst of jobs will scale the cluster to the max when that specific bunch of jobs could have been computed by a single node in the time it took to scale up. I need to be able to describe to the autoscaler the 'weight' of the job so that five hundred tiny jobs cause maybe a single node to be provisioned, vs five hundred massive jobs that actually require a large number of nodes. These are not submitted as batch jobs because they are heterogenous, and submitted by multiple sources at the same time, preventing rolling it into a batch job (the only thing that would give me anyway is the parallelism option, which doesn't fully solve the problem). I would be submitting large portions of these bursts of jobs under the same `max-scale-group` annotation. This will prevent what I have now, which is a large section of recently provisioned nodes that couldn't start before the queue was drained and they sit idle for their entire lifetime. The burst intervals can be just often enough that the unused nodes are recreated a few minutes after being scaled down, and the whole issue repeats.

Another reason I want this is to limit the rate a single entity/submitter can scale the cluster, if they have a lower priority/class of service. The jobs would be assigned relatively large `max-age-before-scaling` and small `max-scale-candidates` values.

**Comment by x13n** [comment]
Got it, thanks. I think this might be a better fit for a dedicated API rather than extra annotations with complicated semantics. Either way, a design doc describing this would be nice. @ahg-g - is this use case something you're looking at?

**Comment by ahg-g** [comment]
Thanks for looping me in.

Regarding the api-server, ideally the API server should be able to protect itself, the hope is that priority and fairness when complete will do that.

Regarding the gradual scale up for jobs, we are generally looking at proper queueing. I see value in what @innovate-invent is proposing, although I feel it will be too complicated for users to figure out what values should be set for each job (unless you want to automate it, in which case this could become a feature of a job queue).

**Comment by innovate-invent** [comment]
The goal is to automatically assign the values based on the user or type of work. This would be done by the application submitting the job. I wouldn't expect (or want to provide access) end users to set these values directly.

**Comment by ahg-g** [comment]
Right, that is why this feels like a queue feature.

**Comment by innovate-invent** [comment]
Ah, I think I misunderstood your previous response. A queue wouldn't fit my requirements. The goal isn't to regulate the rate that work is being submitted to the cluster, I wan't all jobs to execute as soon as resources are available. The goal is to have a mechanism to regulate the rate that the cluster scales up resources. Regulating the rate that jobs are submitted would mean I would need preknowledge of how long the job will actually take. This would also require awkwardly trying to stay ahead of the autoscalers polling interval. I would have to always make sure that there are enough jobs submitted from the queue to saturate the current resources, but not enough to trigger the autoscaler.

I want all jobs to be ready and waiting for the controller to assign to nodes as soon as it finds resources, but have a mechanism that allows me to tell the autoscaler to hold off on scaling the resources for select jobs.

**Comment by ezh** [comment]
FYI MR https://github.com/kubernetes/autoscaler/pull/4258 introduces `node-readiness-grace-time` option which is also could be used to control scale-up speed.

**Comment by damirda** [comment]
@innovate-invent I tried to keep this PR smaller as it is my first contribution to `autoscaler` but once it gets merged, I might implement the second half as well.
